E2E: new `Editor: Post Lifecycle` spec for Playwright.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR proposes a migration of the entire Post Lifecycle to Playwright.

Key changes:
- inclusion of the following steps: `publishing, editing, revert to draft, soft trash, hard delete` of a post.
- addition of methods, objects and components to support new actions.

#### Testing instructions

Related to https://github.com/Automattic/wp-calypso/issues/55950

worldomonation committed:
Migrate Editor: Edit Post spec as a standalone spec.
- new methods in ParagraphBlock to interact with the typography settings.
- new method in GutenbergEditorPage to update the post.
- new page object `PostsPage`.

worldomonation committed:
Add new components or update existing components.
- add new component SnackbarNotificationComponent which handles the toast notifications.
- update EditorSettingsSidebarComponent with method to trash post.

worldomonation committed:
Add or update page objects.
- new object PostsPage.
- add methods in GutenbergEditorPage to handle unpublishing.

worldomonation committed:
Combine specs into one Lifecycle spec.

worldomonation commented:
@dpasque This is a draft of the entire Post Lifecycle spec as outlined in pciE2j-sq-p2.

Out of the original areas of focus (below)
```
– Publishing, editing, reverting to draft, scheduling, trashing, hard deleting
```

I've chosen to implement the following in this spec:
```
– Publishing, editing, reverting to draft, trashing, hard deleting
```

worldomonation commented on packages/calypso-e2e/src/lib/pages/posts-page.ts:
```diff
@@ -0,0 +1,130 @@
+	async clickMenuItem( menuItem: string ): Promise< void > {
+		const locator = this.page.locator( selectors.menuItem( menuItem ) );
+
+		// {@TODO} In the future, a possible idea may be to implement a following structure:
+		// pre-process
+		// perform the menu click
+		// post-process
+		// This is because sometimes the action performed on the menu may require additional
+		// pre- and post-processing, such as in the case of Delete Permanently.
+		// The pre-process and post-process actions are to be called through either a
+		// case-switch statement, or by locating and exeucting predefined function in
+		// an dictionary object, keyed by the value of menuItem.
```
As I mention here, the current `if` clause before and after the actual `click` action works _for now_ because only one action is implemented. 

But in the future, as the list of actions that are implemented grows, some of them will inevitably need some pre- and post-processing like `Delete Permanently`. 

### Why a pre- and post-processing here in the first place?
In deciding on this approach, I wanted to eliminate the requirement for test writers to have to add their own handlers for dialogs, popups, or response waiting, etc. That approach would greatly simplify this `clickMenuItem` method, but shifts the burden of checks onto the test spec itself - and our longstanding approach has been to simplify the specs where possible.

dpasque replied:
As an optional alternative, I think we could "pull up" the processing handlers out of this function to either the `clickMenuItemForPost`, or we could even then make distinct functions for the actions (such as `Delete Permanently`). I think this could work because the pre & post processing involves registering handlers or awaiting promises, all of which can be set up before the call to `clickMenuItem`.  That way, as we need to add special handling in the future, we could continue to add public "top level" functions that include that handling.

Also, from what I can tell, it seems like `Delete Permanently` might be one of the only ones that needs intentional pre- and post- processing? (All the other ones seem to be quick on-page actions, or navigations elsewhere).

dpasque replied:
Here's an example of the proposed above:

```typescript

async deletePostPermanently( title: string ) {
    this.page.once( 'dialog', async ( dialog ) => {
		await dialog.accept();
	} );
	
	await Promise.all( [
	     this.page.waitForResponse( /.*delete.*/ ),
	     this.clickMenuItemForPost( { title: title, action: 'Delete Permanently' } )
	] );

}

async clickMenuItemForPost( {
		title,
		action,
	}: {
		title: string;
		action: MenuItems;
	} ): Promise< void > {
		await this.ensurePostShown( title );

		await this.togglePostMenu( title );
		await this.clickMenuItem( action );
	}

private async clickMenuItem( menuItem: string ): Promise< void > {
		const locator = this.page.locator( selectors.menuItem( menuItem ) );
        await locator.click();
}


```

worldomonation replied:
I haven't checked through the entire list of actions possible spanning across the tabs (Post, Draft, Scheduled, Trashed). 

My reason for staying away from single-purpose methods (eg. `deletePostPermanently`) was avoid the pitfalls of the Selenium framework where the actions were nearly all implemented with single-purpose methods. I do recognize however that there is a time and place for them, and this might be one of them!

Now, if the `Delete Permanently` is indeed the only action that requires pre and post-processing handlers, then I am satisfied keeping the current structure for now. 

I'll explore ways to combine your suggestion with keeping a unified interface.

worldomonation commented on packages/calypso-e2e/src/lib/blocks/paragraph-block.ts:
```diff
@@ -1,7 +1,7 @@
 
 const selectors = {
-	block: '.wp-block-paragraph',
+	block: 'p',
```
Previously this did not cause a problem because we never used `ParagraphBlock`'s `validate` method.

dpasque replied:
Just having the `p` here seems a little strange to me. At this point, could we just validate the text we entered is present with a plain text selector? That should simplify it a bit and be just about as reliable.

worldomonation replied:
We do the check later on in the `validatePostContent` method - or do you mean a validation while in the editor view?

dpasque replied:
Oh I meant that I think you could change the selector used in the `validatePublishedContent` to not reference the selector here (since it just adds a `p`) and just look at the text itself alone (without the `p`). But, completely optional 🙂

worldomonation commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -36,6 +36,8 @@ const selectors = {
 	publishButton: ( parentSelector: string ) =>
 		`${ parentSelector } button:text("Publish")[aria-disabled=false]`,
+	updateButton: 'button:text("Update")',
+	switchToDraftButton: 'button.editor-post-switch-to-draft',
```
The button that handles saving/unpublishing (commonly referred to as the draft button) has different selectors based on the state of the post, hence this new selector to locate the button when the post is in an unpublishable state.

dpasque replied:
Good call!

worldomonation committed:
Create new method `ensurePostShown` that is a superset of the `waitUntilLoaded`.

worldomonation committed:
Use locator for `waitForPostToAppear` instead

dpasque commented on packages/calypso-e2e/src/lib/components/snackbar-notification-component.ts:
```diff
@@ -0,0 +1,52 @@
+	 * @returns {Promise<boolean>} True if text is found in target snackbar notice.
+	 * 	False otherwise.
+	 */
+	async validateNoticeShown(
```
I really like this style of checking for the snackbars!

Small stylistic nit: could we change the function name to `noticeIsShown` because it returns a boolean?

Throughout our tests, `validate*` functions usually return void and throw an error on failure. This doesn't do that, but returns a boolean, so I think it might be good to make the function name indicate that.

dpasque commented on packages/calypso-e2e/src/lib/components/editor-settings-sidebar-component.ts:
```diff
@@ -130,6 +131,16 @@ export class EditorSettingsSidebarComponent {
+	/**
+	 * Trashes the current post.
+	 */
+	async trashPost(): Promise< void > {
+		await Promise.all( [
+			this.page.waitForNavigation(),
+			this.frame.click( selectors.trashPostButton ),
+		] );
+	}
```
I realize this might involve some spec redesign and so may not be worth it, but what do you think of trashing from the `/posts` page?

I was thinking from the "put Calypso under test" perspective -- we make a unique UI for the `/posts` page, and as much as we can test that page as opposed to the editor iframe might be a good idea (since the editor already has some tests in core).

What do you think?

worldomonation replied:
Ah, good point - I had thought the trash from editor and the subsequent redirect to the PostsPage were a Calypso specific behavior, but checked on JurassicNinja and turns out that's just how WordPress operates!

I'll revise this section.

dpasque commented on packages/calypso-e2e/src/lib/pages/posts-page.ts:
```diff
@@ -0,0 +1,166 @@
+		 * @param {Page} page Page object.
+		 */
+		async function waitForPostToAppear( page: Page ): Promise< void > {
+			const postLocator = page.locator( selectors.postItem( title as string ) );
```
I might be missing the use case, but I don't think you need the type assertion here? The `as string` piece, that is.

worldomonation replied:
Oops, that was left over from when I was exploring ways to perform post visible assertion in `waitUntilLoaded`. Thanks for the catch.

dpasque commented on packages/calypso-e2e/src/lib/pages/posts-page.ts:
```diff
@@ -0,0 +1,166 @@
+			await postLocator.waitFor( { state: 'visible', timeout: 20 * 1000 } );
+		}
+
+		await reloadAndRetry( this.page, waitForPostToAppear );
```
This is a great idea for handling the race condition! 🎉

dpasque commented on packages/calypso-e2e/src/lib/pages/posts-page.ts:
```diff
@@ -0,0 +1,166 @@
+
+		if ( menuItem === 'Delete Permanently' ) {
+			await this.page.waitForResponse( /.*delete.*/ );
+		}
```
I'm not entire sure this is accurate, but since it's common use case to wait for the snackbar after this, can we skip having to wait on the web request? Or could we pull the wait for the snackbar right into this function instead of the web request wait?

worldomonation replied:
Hmm, I thought the `delete permanently` didn't produce a snackbar, but apparently it does! I will try switching over to waiting for the snackbar toast.

worldomonation replied:
Doh 🤦 I even codified the snackbar wait in the spec itself, so I'm just forgetting things..

dpasque commented on test/e2e/specs/specs-wpcom/wp-editor__post-advanced-flow.ts:
```diff
@@ -0,0 +1,162 @@
+		const testAccount = new TestAccount( accountName );
+		await testAccount.authenticate( page );
+
+		gutenbergEditorPage = new GutenbergEditorPage( page );
+		await gutenbergEditorPage.visit( 'post' );
```
Also in the vein of "let's test more Calypso", what do you think of starting this test on the `/posts` page and clicking on the "Add new post" button? Might be a good validation that takes you into the editor correctly.

worldomonation replied:
Interesting call. I'll tinker with it.

dpasque reviewed (commented):
This is arguably of the hardest tests to design and implement in the wpcom transition, and you did a fantastic job. This reads really cleanly and has awesome coverage! Nice work! 😄 

Have a few stylistic and test design questions -- definitely not married to any of them, so let me know what you think!

worldomonation committed:
Revise PR taking in suggestions:
- trash post from the PostsPage.
- remove unnecessary handling from Delete Permanently case.
- remove unnecessary string assertion.
- rename snackbar method.

dpasque reviewed (approved):
Looks like some of the builds are broken, but the code changes look great! Approving for once all the builds get cleaned up! Nice work 😄

worldomonation committed:
Make the clickTab method more robust:
- wait for `networkidle`.
- use locators for the `ElementHelper.clickNavTab` method.
